Panel scrollbar update

KenediaDev opened the pull request.

KenediaDev committed:
First Draft

KenediaDev committed:
Resizing Loop fix

- Reflow functions now use the *ContentRegion* Width instead of the Width of the Panel
- Default Scrollbar Padding is 4 px
- to handle panels inside panels which use FILL to be sized correctly some scrollbar properties have been moved up from (Flow)Panel to Container
- Content Regions only resize to their VISIBILE childs size instead of accounting for any which resulted in panels to not grow back to smaller sizes when hiding bigger controls
- Scrollbar now returns the _barBounds.Width instead of the _textureBar.Width

KenediaDev committed:
added Max Size to Containers

KenediaDev committed:
Fixed a logic error

BillChirico commented on Blish HUD/Controls/Container.cs:
```diff
@@ -146,6 +196,7 @@ public abstract class Container : Control, IEnumerable<Control> {
 
         public bool AddChild(Control child) {
             if (_children.Contains(child)) return true;
+            if (_panelScrollbar != null && child == _panelScrollbar) return true;
```
Can be shortened to

`return _panelScrollbar != null && child == _panelScrollbar`

KenediaDev replied:
Wouldn't that return always and not continue to add the child?

dlamkins committed:
Merge branch 'dev' into Panel-Scrollbar-Update

dlamkins commented:
Can you please include a description of all of the changes or provide a link to a full summary of where you've described it in Discord?

dlamkins commented:
![image](https://user-images.githubusercontent.com/1950594/188567510-a5b4d211-727f-4faf-9ee8-c00f7354950a.png)

I'm seeing weird issues with many of the containers.  This is a bit confusing with how it appears to be modifying the content region.

KenediaDev committed:
Removed changes caused by formatting

KenediaDev commented:
### Container.cs

Properties:
- added a _panelScrollbar property which is either null or the scrollbar of the container
- added a new public property ScrollPadding which determines the Padding on the left side of the Scrollbar
- added a new public bool property ScrollbarVisible which tells if the scrollbar is visible or not
- added a new public int property ScrollbarWidth which returns the current scrollbar width (0 if null)
- added a new public Point property MaxSize which lets you limit the size a container would get resized to
- added a new public int property MaxWidth which lets you limit the width a container would get resized to -> MaxSize.x
- added a new public int property MaxHeight which lets you limit the width a container would get resized to -> MaxSize.y

Methods:
- exclude the scrollbar from the Child handling of the container
- take MaxSize into account when resizing the container
- disposing the scrollbar on dispose


____________________

### Scrollbar.cs

Properties:
- added a new public bool property Drawn which tells if the scrollbar is visible and drawn or not
- added a new public int property ScrollbarWidth which returns the scrollbar width

____________________

### Panel.cs

Methods:
**RecalculateLayout**
- updating the ScrollbarVisible property
- updating the _panelScrollbar if not null to match the container size

**UpdateScrollbar**
- adjusted _panelScrollbar.Right to get the position right
	
____________________

### FlowPanel.cs

Methods:
**all variants of ReflowChildLayout**
- using the ContentRegion instead of the container bounds for layout calculations to avoid the scrollbar bounds

KenediaDev commented:
moved to https://github.com/blish-hud/Blish-HUD/pull/749 for a cleaner PR
